Release notes and migration guide ready to merge (#2370)
tinymce/tinymce-docs: modules/ROOT/pages/6.0-release-notes-events.adoc

= TinyMCE 6.0 Event changes
:navtitle: TinyMCE 6.0 Event changes
:description: TinyMCE 6.0 Event changes
:keywords: releasenotes, events

This section documents the improvements and fixes made to {productname} 6.0 Events.

* xref:events[Events]

> * xref:new-and-improved-events][New and improved Events]

Reviewer: ```suggestion
* xref:new-and-improved-events[New and improved Events]
```
This xref is broken atm.

* xref:fixed-events[Fixed Events]

// tag::events[]
[[events]]
== Events

[[new-and-improved-events]]
=== New and improved events

[[autocompleter]]
==== `AutocompleterStart`, `AutocompleterUpdate` and `AutocompleterEnd`

> As part of moving `autocompleter` from the default `silver` theme to the {productname} core, three new events are now available: `AutocompleterStart`, `AutocompleterUpdate` and `AutocompleterEnd`.

Reviewer: Yeah I agree, we need to make this more specific. The suggestion isn't correct though, as it's not the UI that was moved to core. Instead, it was the core autocompleter logic (e.g keyboard event hooks, or everything except the UI). Maybe something like this?
```suggestion
As part of moving the core of the autocompleter functionality from the default `silver` theme to the {productname} core, three new events are now available: `AutocompleterStart`, `AutocompleterUpdate` and `AutocompleterEnd`.
```

> Themes can act on the `AutocompleterStart` and `AutocompleterUpdate` events and can call `AutocompleterEnd` to cancel the auto completer.

Reviewer: Another option would be to link to the editor themes page which would provide the extra context?

> Themes can act on the `AutocompleterStart` and `AutocompleterUpdate` events and can call `AutocompleterEnd` to cancel the auto completer.

Reviewer 2: Deferring

[[before-set-content]]
==== `BeforeSetContent`

> The `BeforeSetContent` event now includes the actual serialized content when passing in an `AstNode` to the `editor.setContent`.

Reviewer: ```suggestion
The `BeforeSetContent` event now includes the actual serialized content when passing in an `AstNode` to the `editor.setContent()` API.
```

[[get-content]]
==== `GetContent`

The `GetContent` event now always passes a `string` for the `content` property.

[[theme-load-error]]
==== `ThemeLoadError`

A `ThemeLoadError` event is now fired if the theme fails to load.

[[fixed-events]]
=== Fixed events

[[before-input]]
==== `beforeinput`

The `beforeinput` event is expected to fire whenever material is added to or removed from a {productname} instance.

> Previously, however, because of overrides in the `DeleteBackspaceKeys.ts` module, the `beforeinput` event did not always fire when expected.

Reviewer: These are internal files so not something we should be mentioning to customers. Maybe something like this instead?
```suggestion
Previously, however, because of backspace and delete overrides in {productname} Core, the `beforeinput` event did not always fire when expected.
```

> With this release, these over-rides no longer prevent `beforeinput` events firing when expected.

Reviewer 2: ```suggestion
With this release, these overrides no longer prevent `beforeinput` events firing when expected.
```
Consistency with above sentence

[[change]]
==== `change`

Previously, the `change` event was triggered by the first modification made to a {productname} editor instance.

However, subsequent changes did not cause the event to fire until focus was switched away from the editor.

For example, typing a single character in to the {productname} editor text-entry area triggered the `change` event.

Consequent typing, or selecting of the entered character and modifying it by setting it to *bold* or _italic_, did not trigger further `change` events.

Making further modifications and then switching focus away from the editor did trigger this event, however.

As of {productname} 6, the `change` event is not fired unless and until

> . focus is switched away from the editor; and

Reviewer 3: ```suggestion
. Focus is switched away from the editor; and
```

Author: The two bullet point items are clauses, each one part of the complete sentence started by the text immediately above: *…the `change` event is not fired unless and until*. Consequently, the two clauses don’t take an initial Capital.

> . changes have been made in the editor since focus was switched to it.

Reviewer 3: ```suggestion
. Changes have been made in the editor since focus was switched to it.
```

Author: The two bullet point items are clauses, each one part of the complete sentence started by the text immediately above: *…the `change` event is not fired unless and until*. Consequently, the two clauses don’t take an initial Capital.

> NOTE: it is still possible to listen for modification to a {productname} editor instance before focus is changed. To do this, listen for the `dirty` event, rather than the `change` event.

Reviewer 2: ```suggestion
NOTE: It is still possible to listen for modification to a {productname} editor instance before focus is changed. To do this, listen for the `dirty` event, rather than the `change` event.
```

[[fixed-get-content]]
==== `GetContent`

Previously the `GetContent` event did not fire when the `editor.selection.setContent()` API attempted to get `tree` or `text` formatted data.

> With this release, the `GetContent` event fires as expected when `editor.selection.setContent()` attempts this.

Reviewer: ```suggestion
With this release, the `SetContent` event fires as expected when `editor.selection.setContent()` attempts this.
```

> NOTE: The `editor.selection.setContent()` API only supports the `html` and `raw` formats. It does not support `tree` or `text` formats. To create a tree to pass in to `editor.selection.setContent()`, use the `new tinymce.html.Node()` API, or parse to a tree using `editor.parser.parse(html)`.

Reviewer: This doesn't make sense? We say it doesn't support the `tree` format, but then we tell them to create one and use it with the API? I think we should just drop the last sentence entirely.

[[input]]

> ==== `input`

Reviewer: This has a direct correlation to `beforeinput` and they are part of the same event cycle. Having these separate to me makes this more confusing so I think we should merge it into the one detail.

> The `input` event is expected to fire whenever material is deleted from a {productname} instance.

Reviewer 2: ```suggestion
The `input` event is expected to fire whenever content is deleted from a {productname} instance.
```
As above - more our usual wording?

> Previously, however, because of over-rides in the `DeleteBackspaceKeys.ts` module, when more than one line was deleted, the `input` event did not fire.

Reviewer: If this section isn't merged with `beforeinput` then likewise here we shouldn't mention `DeleteBackspaceKeys.ts` as it's an internal implementation detail.
